File: rag/src/chunker.js
/**
 * chunker.js — 语义切块器
 *
 * 职责：
 * - 基于标题层级（Markdown headings）切分 chunk
 * - 列表、表格、规则块尽量独立
 * - 每个 chunk 只讲一个核心点
 * - 单块 300-900 汉字
 *
 * chunk_type 枚举：
 *   definition, rule, checklist, procedure, anti_pattern, template, example, navigation
 */

/**
 * 检测 chunk 类型
 */
function detectChunkType(section) {
  const lines = section.split('\n');
  const text = section.toLowerCase();
  const firstHeading = lines.find(l => l.startsWith('#')) || '';

  // 导航/索引
  if (/索引|导航|目录|index/i.test(firstHeading) &&
      (text.includes('├─') || text.includes('└─') || text.includes('|-') || text.includes('---'))) {
    return 'navigation';
  }

  // 模板/选型
  if (text.includes('项目选型') || text.includes('模板') ||
      text.includes('基本信息') || text.includes('| 字段 |')) {
    return 'template';
  }

  // 红线/禁止
  if (text.includes('红线') || text.includes('禁止') || text.includes('不要') ||
      text.includes('不允许') || text.includes('禁区') || text.includes('毒点')) {
    return 'rule';
  }

  // 步骤/流程
  if (text.includes('step') || text.includes('步骤') || text.includes('顺序') ||
      text.includes('第') && text.includes('步')) {
    return 'procedure';
  }

  // 清单/自检
  if (text.includes('检查') || text.includes('清单') || text.includes('自检') ||
      text.includes('评估') || text.includes('检测')) {
    return 'checklist';
  }

  // 反模式
  if (text.includes('常见问题') || text.includes('常见错误') || text.includes('不要') ||
      text.includes('反例') || text.includes('忌讳')) {
    return 'anti_pattern';
  }

  // 示例
  if (text.includes('示例') || text.includes('例子') || text.includes('比如') ||
      text.includes('例如')) {
    return 'example';
  }

  // 定义/原则
  if (text.includes('定义') || text.includes('原则') || text.includes('概述') ||
      text.includes('概念') || text.includes('范围')) {
    return 'definition';
  }

  return 'definition';
}

/**
 * 清理文本：移除多余空行、提取纯文本
 */
function cleanText(text) {
  return text
    .split('\n')
    .map(l => l.trim())
    .filter(l => l.length > 0)
    .join('\n');
}

/**
 * 按标题切分 Markdown 内容
 * @param {string} content - 原始 markdown
 * @returns {Array<{title: string, text: string, heading_level: number, section_ref: string}>}
 */
export function chunkMarkdown(content) {
  const lines = content.split('\n');
  const sections = [];
  let currentSection = null;
  let sectionCounter = 0;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const headingMatch = line.match(/^(#{1,6})\s+(.+)/);

    if (headingMatch) {
      // 保存上一节
      if (currentSection && currentSection.text.length > 5) {
        sections.push(currentSection);
      }

      const level = headingMatch[1].length;
      const title = headingMatch[2].trim();
      sectionCounter++;

      const sectionRef = title
        .toLowerCase()
        .replace(/[^\w一-鿿]+/g, '-')
        .replace(/-+/g, '-')
        .replace(/^-|-$/g, '')
        .substring(0, 60);

      currentSection = {
        title,
        text: '',
        heading_level: level,
        section_ref: sectionRef || `section-${sectionCounter}`
      };
    } else if (currentSection) {
      currentSection.text += (currentSection.text ? '\n' : '') + line;
    } else {
      // 文档开头 (# 之前的内容)
      if (!currentSection) {
        currentSection = {
          title: '文档开头',
          text: '',
          heading_level: 1,
          section_ref: 'doc-intro'
        };
      }
      currentSection.text += (currentSection.text ? '\n' : '') + line;
    }
  }

  // 保存最后一节
  if (currentSection && currentSection.text.length > 5) {
    sections.push(currentSection);
  }

  // 后处理：将过长的 section 再按子标题/序号拆分
  const refined = [];
  for (const section of sections) {
    const charCount = countChars(section.text);
    if (charCount <= 900) {
      refined.push(section);
      continue;
    }

    // 尝试按子结构拆分
    const subSections = splitLongSection(section);
    refined.push(...subSections);
  }

  // 为每个 section 计算 chunk_type
  return refined.map(s => ({
    ...s,
    chunk_type: detectChunkType(s.title + '\n' + s.text),
    text: cleanText(s.text)
  })).filter(s => s.text.length >= 10); // 过滤太短的块
}

/**
 * 计算中文字符数
 */
function countChars(text) {
  // 中文算 1 个，英文/数字算 0.5，折合汉字数
  let count = 0;
  for (const ch of text) {
    if (/[一-鿿]/.test(ch)) count += 1;
    else if (/\S/.test(ch)) count += 0.3;
  }
  return count;
}

/**
 * 拆分过长的 section
 */
function splitLongSection(section) {
  const lines = section.text.split('\n');
  const subSections = [];
  let current = { ...section, text: '', section_ref: section.section_ref + '-a' };
  let subIdx = 'abcdefghij'.split('');

  for (const line of lines) {
    // 尝试在序号项处拆分
    const isSubItem = /^(\d+[\.\)、]|[-*•]\s)/.test(line.trim());
    const isTableRow = /^\|/.test(line.trim());

    if (isSubItem && countChars(current.text) > 300) {
      subSections.push({ ...current, text: cleanText(current.text) });
      const idx = subIdx.shift() || 'x';
      current = {
        ...section,
        title: `${section.title} (${idx})`,
        text: line,
        section_ref: section.section_ref + '-' + idx
      };
    } else {
      current.text += '\n' + line;
    }
  }

  if (current.text.trim().length > 5) {
    subSections.push({ ...current, text: cleanText(current.text) });
  }

  return subSections.filter(s => s.text.length >= 10);
}

export default { chunkMarkdown };

File: rag/src/chunker.py
"""
chunker.py — Markdown 文档分块器

职责：
- 按 heading 级别（#, ##, ###）做语义边界分割
- 检测 chunk_type（definition/rule/checklist/procedure/anti_pattern/template/example/navigation）
- 控制块长度在 300-900 字之间
- 超出 900 字的块在子项处分割
- 输出统一的 chunk 结构
"""

import re
from pathlib import Path


def _detect_chunk_type(text, heading_level):
    """检测块类型"""
    lower = text.lower()

    # 导航/索引
    if heading_level >= 2 and (
        re.search(r"^[\s]*(?:目录|index|导航|概览|总览|一览)", text, re.MULTILINE)
        or re.search(r"(?:^|\n)(?:[\d.]+\s*[-—]\s*.+){5,}", text)
        and len(text) < 300
    ):
        return "navigation"

    # 模板
    if re.search(r"(?:模板|示例|例子|范例|示例：)", lower) and re.search(
        r"\{|【|\[.*\]", text
    ):
        return "template"

    # 规则
    if re.search(r"(?:^|\n)\s*(?:规则|原则|规范|必须|不要|禁止|避免|可以|建议)", lower):
        return "rule"

    # 步骤/流程
    if re.search(r"(?:^|\n)\s*(?:步骤|流程|方法|做法|如何|操作|方式|过程)", lower) or (
        re.search(r"(?:首先|第一步|第二步|最后|然后)", lower)
        and re.search(r"\d+[.、]", text)
    ):
        return "procedure"

    # 检查清单
    if re.search(r"(?:^|\n)\s*(?:清单|检查|检查点|核对|审核|验收)", lower) or re.search(
        r"(?:^|\n)\s*[-*]\s*\[[\sx]\]", text
    ):
        return "checklist"

    # 反模式
    if re.search(r"(?:反模式|禁忌|陷阱|误区|避免|不要)", lower) and re.search(
        r"(?:^|\n)\s*(?:不要|避免|禁忌|陷阱)", lower
    ):
        return "anti_pattern"

    # 示例
    if re.search(r"(?:^|\n)\s*(?:举例|示例|例如|比如|譬如)", lower):
        return "example"

    return "definition"


def _split_long_section(text, max_chars=900):
    """将过长的文本在子项处分割

    参数：
        text: 段文本
        max_chars: 最大字符数

    返回：
        字符串列表
    """
    if len(text) <= max_chars:
        return [text]

    chunks = []
    # 尝试在 markdown 列表项、空行或句号处分割
    lines = text.split("\n")
    current = []

    for line in lines:
        current_len = sum(len(l) for l in current) + len(current) - 1  # 换行符数量

        if current_len + len(line) > max_chars and current:
            # 在列表项处分割
            content = "\n".join(current)
            if content.strip():
                chunks.append(content.strip())

            # 保留列表项前缀（如果是列表继续项）
            stripped = line.lstrip()
            if stripped.startswith("-") or stripped.startswith("*") or re.match(r"^\d+[.、]", stripped):
                current = [line]
            else:
                current = [line]
        else:
            current.append(line)

    if current:
        content = "\n".join(current)
        if content.strip():
            chunks.append(content.strip())

    return chunks if chunks else [text]


def _clean_text(text):
    """清理文本，去掉多余空行"""
    # 去掉开头和结尾的空白
    text = text.strip()
    # 将连续 3 个以上空行压缩为 2 个
    text = re.sub(r"\n{4,}", "\n\n\n", text)
    return text


def chunk_markdown(document):
    """将 document 按 heading 分割为 chunks

    参数：
        document: normalize_document 返回的字典（含 _raw_content）

    返回：
        chunk 字典列表
    """
    content = document.get("_raw_content", "")
    if not content:
        return []

    chunks = []
    lines = content.split("\n")
    current_section_lines = []
    current_heading = document["title"]
    current_level = 1

    def flush_section():
        """把当前累积的 section 输出为一个或多个 chunk"""
        nonlocal current_section_lines
        if not current_section_lines:
            return

        text = "\n".join(current_section_lines).strip()
        if not text:
            return

        chunk_type = _detect_chunk_type(text, current_level)

        # 如果文本过长，在子项处分割
        sub_sections = _split_long_section(text, max_chars=900)

        for i, sub_text in enumerate(sub_sections):
            sub_text = _clean_text(sub_text)
            if len(sub_text) < 20:
                continue

            chunk_id = f"{document['doc_id']}.sec{len(chunks) + 1}"
            if i > 0:
                chunk_id += f".{i + 1}"

            chunks.append({
                "chunk_id": chunk_id,
                "doc_id": document["doc_id"],
                "title": current_heading,
                "chunk_type": chunk_type,
                "text": sub_text,
                "category": document["category"],
                "stage": document["stage"],
                "scope": document["scope"],
                "tags": document["tags"],
                "priority": document["priority"],
                "source_type": document["source_type"],
                "source_path": document["source_path"],
                "heading_level": current_level,
            })

        current_section_lines = []

    for line in lines:
        heading_match = re.match(r"^(#{1,3})\s+(.+)$", line)
        if heading_match:
            # flush 上一段
            flush_section()

            level = len(heading_match.group(1))
            heading_text = heading_match.group(2).strip()

            # 跳过标题行自身（它已经被识别为分割标记）
            current_level = level
            current_heading = heading_text
            # 不把标题行本身加入内容
            continue

        current_section_lines.append(line)

    # flush 最后一段
    flush_section()

    # 如果没分出 chunk（可能整个文档没有 heading），则整个作为一块
    if not chunks and content.strip():
        text = _clean_text(content)
        if len(text) >= 20:
            chunks.append({
                "chunk_id": f"{document['doc_id']}.sec1",
                "doc_id": document["doc_id"],
                "title": document["title"],
                "chunk_type": _detect_chunk_type(text, 1),
                "text": text,
                "category": document["category"],
                "stage": document["stage"],
                "scope": document["scope"],
                "tags": document["tags"],
                "priority": document["priority"],
                "source_type": document["source_type"],
                "source_path": document["source_path"],
                "heading_level": 1,
            })

    return chunks
